cmd/ipfs/main.go

```diff
@@ -315,6 +315,13 @@ func callCommand(ctx context.Context, req cmds.Request, root *cmds.Command) (cmd
 		return nil, err
 	}
 
+	if cmd.PreRun != nil {
+		err = cmd.PreRun(req)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
might want to make `PreRun` and `PostRun` the same signature. i.e. returning an err in one is a bit odd. maybe it can also be called with `(req, res)` and use `res.SetError(.)`, and right below here we can check if prerun set an error and fail early.